docs/latex/Inhalt/03_Inhalt/04_unit_tests.tex

```diff
@@ -0,0 +1,33 @@
+
+Für die Peripherie der Software wurde zunächst auf Unit Tests verzichtet, da dieser Teil des Programms ohnehin leicht und häufig austauschbar sein soll und daher von geringerer Wichtigkeit ist, sodass ihm auch beim Testen eine geringere Priorität beigemessen wurde. Lediglich in der Adapterschicht wird eine Klasse, der \code{OpinionMapper} getestet. Auch für den Domaincode wurde zunächst auf Unit Tests verzichtet, da sich die dort implementierte Businesslogik hauptsächlich auf die Definition von Klassen und ihre Attribute beschränkt, es existieren jedoch keine Methoden, welche tatsächlich komplexere Logik enthalten.
+
+\section{ATRIP-Regeln}
+Bei den ATRIP-Regeln handelt es sich um fünf grundlegende Regeln für die Erstellung guter Unit Tests. Inwiefern diese für das vorliegende Projekt eingehalten wurden, wird in den nachfolgen Abschnitten aufgezeigt.
+
+\paragraph{Automatic} Alle implementierten Unit Tests laufen vollständig eigenständig ab, da keinerlei manuelle Eingriffe, etwa in Form von Werteeingaben, notwendig sind. Außerdem überprüfen alle Tests ihre Ergebnisse durch Assertions automatisch.
+
+\paragraph{Thorough} Da es sich bei dieser Regel um eine \enquote{weiche} Regel handelt, ist eine eindeutige Entscheidung, ob alles Notwendige getestet wurde und damit die Regel erfüllt ist schwierig. Als \enquote{notwendig} wurde für das vorliegende Projekt die wesentliche Businesslogik und damit die Applikationsschicht definiert. Diese wird mit einer hohen Code-Coverage getestet, sodass die implementierten Tests durchaus als gründlich bezeichnet werden können. (vgl. \autoref{sec:code-coverage}) Es wurden allerdings keine Tests für spezifische Softwarefehler implementiert, da sich die Software zu dem Zeitpunkt, da dieses Dokument verfasst wird, noch in der Entwicklung befindet und daher bisher noch keine Softwarefehler gemeldet wurden.
+
+\paragraph{Repeatable} Die implementierten Unit Tests sind jederzeit automatisch durchführbar und liefern das gleiche Ergebnis, da sie weder zeit- noch zufallsabhängige Komponenten beinhalten und auch keine Abhängigkeiten auf Dateisysteme, Datenbanken oder ähnliches besitzen bzw. diese durch Mocks ersetzt werden, welche stets die gleichen Daten liefern. (vgl. \autoref{sec:mocks})
+
+\paragraph{Independent} Die implementierten Unit Tests sind jederzeit in beliebiger Reihenfolge ausführbar. Dies wird sichergestellt, indem alle Tests strikt nach der AAA-Normalform aufgebaut sind, also jeder Test in seiner ersten Phase seine eigene \enquote{Testwelt} initialisiert. Dabei werden jegliche gewöhnlich persistente Daten durch Mocks bereitgestellt, welche ebenfalls in der ersten Phase jedes Tests neu trainiert werden, sodass auch keine impliziten Abhängigkeiten auf bestehende Daten entstehen können.
```

Comment:
"Dabei werden jegliche gewöhnlich persistente Daten durch Mocks bereitgestellt" => Ich verstehe den ersten Teil des Satzteils nicht:
1. Es wird nicht nur die Persistenz weggemockt (aber hier ist nur das relevant?)
2. Ich verstehe die Formulierung nicht

Reply:
Ich hab mal das "gewöhnlich" weggeschmissen, vllt wird es dann klarer. Die Idee ist: Hätten wir an der Stelle eine richtige Persistenz, können wir implizite Abhängigkeiten über Daten haben. Aber die ersetzen wir ja durch Mocks, die in jedem Test neu trainiert werden. Es kann also hierüber keine Abhängigkeiten geben.

Reply:
Das würde ich vielleicht noch erklären, also "Im Produktivbetrieb gibt es Abhängigkeiten auf persistierte Daten. Um dieses bei Unit Tests zu verhindern, werden jegliche Persistenzzugriffe auf Mocks ausgeführt" oder so